src/brod_utils.erl

```diff
@@ -478,6 +522,52 @@ get_stable_offset(Header) ->
 
 %%%_* Internal functions =======================================================
 
+do_fold(Spawn, {Pid, Mref}, Offset, Acc, Fun, End, Count) ->
+  receive
+    {'DOWN', Mref, process, Pid, Result} ->
+      handle_fetch_rsp(Spawn, Result, Offset, Acc, Fun, End, Count)
+  end.
+
+handle_fetch_rsp(_Spawn, {error, Reason}, Offset, Acc, _Fun, _, _) ->
+  ?BROD_FOLD_RET(Acc, Offset, {fetch_failure, Reason});
+handle_fetch_rsp(_Spawn, {ok, {StableOffset, []}}, Offset, Acc, _Fun,
+                _End, _Count) when Offset >= StableOffset ->
+  ?BROD_FOLD_RET(Acc, Offset, reached_end_of_partition);
+handle_fetch_rsp(Spawn, {ok, {_StableOffset, Msgs}}, Offset, Acc, Fun,
+                 End, Count) ->
+  #kafka_message{offset = LastOffset} = lists:last(Msgs),
+  %% start fetching the next batch if not stopping at current
+  Fetcher = case LastOffset < End andalso length(Msgs) < Count of
+              true -> Spawn(LastOffset + 1);
+              false -> undefined
+            end,
+  do_acc(Spawn, Fetcher, Offset, Acc, Fun, Msgs, End, Count).
+
+do_acc(_Spawn, Fetcher, Offset, Acc, _Fun, _, _End, 0) ->
+  undefined = Fetcher, %% assert
+  ?BROD_FOLD_RET(Acc, Offset, reached_message_count_limit);
+do_acc(_Spawn, Fetcher, Offset, Acc, _Fun, _, End, _Count) when Offset > End ->
+  undefined = Fetcher, %% assert
+  ?BROD_FOLD_RET(Acc, Offset, reached_target_offset);
+do_acc(Spawn, Fetcher, Offset, Acc, Fun, [], End, Count) ->
+  do_fold(Spawn, Fetcher, Offset, Acc, Fun, End, Count);
+do_acc(Spawn, Fetcher, Offset, Acc, Fun, [Msg | Rest], End, Count) ->
+  case Fun(Msg, Acc) of
+    {ok, NewAcc} ->
+      NextOffset = Msg#kafka_message.offset + 1,
+      do_acc(Spawn, Fetcher, NextOffset, NewAcc, Fun, Rest, End, Count - 1);
+    {error, Reason} ->
+      ok = kill_fetcher(Fetcher),
+      ?BROD_FOLD_RET(Acc, Offset, Reason)
+  end.
+
+kill_fetcher({Pid, Mref}) ->
+  exit(Pid, kill),
+  receive
+    {'DOWN', Mref, process, _, _} ->
```

Comment:
Looks like this may hang forever if Pid crashed before. Add `after` clause?

Reply:
You mean raise an exception in the `after` clause?
The pid is from `spawned_monitor`, there is no reason to miss a `DOWN` message unless eaten by the fold callback.